src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -47,72 +56,166 @@ public static StreamingClientProvider getStreamingClientProviderInstance() {
+      resultClient = this.streamingClientHandler.createClient(clientProperties);
+      LOGGER.info(
+          "Streaming client optimization is disabled, creating a new streaming client with name:"
+              + " {}",
+          resultClient.getName());
     }
+
+    return resultClient;
   }
 
   /**
-   * Closes the given client
+   * Closes the given client and deregisters it from the cache if necessary. It will also call close
+   * on the registered client, which should be the same as the given client so the call will no-op.
    *
+   * @param connectorConfig The configuration to deregister from the cache
    * @param client The client to be closed
    */
-  public void closeClient(SnowflakeStreamingIngestClient client) {
-    this.providerLock.lock();
+  public void closeClient(
+      Map<String, String> connectorConfig, SnowflakeStreamingIngestClient client) {
+    StreamingClientProperties clientProperties = new StreamingClientProperties(connectorConfig);
+
+    // invalidate cache
+    SnowflakeStreamingIngestClient registeredClient =
+        this.registeredClients.getIfPresent(clientProperties);
+    if (registeredClient != null) {
+      // invalidations are processed on the next get or in the background, so we still need to close
+      // the client here
+      this.registeredClients.invalidate(clientProperties);
+      this.streamingClientHandler.closeClient(registeredClient);
+    }
+
+    // also close given client in case it is different from registered client. this should no-op if
+    // it is already closed
     this.streamingClientHandler.closeClient(client);
-    this.providerLock.unlock();
+  }
+
+  // TEST ONLY - return the current state of the registered clients
+  @VisibleForTesting
+  public Map<StreamingClientProperties, SnowflakeStreamingIngestClient> getRegisteredClients() {
+    return this.registeredClients.asMap();
+  }
+
+  public static class StreamingClientProperties {
+    public final Properties clientProperties;
+    public final String clientName;
+    public final Map<String, Object> parameterOverrides;
+
+    public StreamingClientProperties(Map<String, String> connectorConfig) {
+      this.clientProperties = StreamingUtils.convertConfigForStreamingClient(connectorConfig);
+
+      this.clientName =
+          STREAMING_CLIENT_PREFIX_NAME + connectorConfig.getOrDefault(Utils.NAME, TEST_CLIENT_NAME);
+
+      // Override only if bdec version is explicitly set in config, default to the version set
+      // inside Ingest SDK
+      this.parameterOverrides = new HashMap<>();
+      Optional<String> snowpipeStreamingBdecVersion =
+          Optional.ofNullable(connectorConfig.get(SNOWPIPE_STREAMING_FILE_VERSION));
+      snowpipeStreamingBdecVersion.ifPresent(
+          overriddenValue -> {
+            LOGGER.info("Config is overridden for {} ", SNOWPIPE_STREAMING_FILE_VERSION);
+            parameterOverrides.put(BLOB_FORMAT_VERSION, overriddenValue);
+          });
```

Comment:
not needed now but I think we can remove this since this is deprecated on server side. 
WDYT @sfc-gh-tzhang